core/src/sumeragi/fault.rs

```diff
@@ -38,866 +37,1337 @@ impl FaultInjection for NoFault {
+    /// This transaction cache was therefore introduced so that only
+    /// interact with the transaction queue as necessary. If this were
+    /// written in C it would have just been an array of
+    /// `VersionedAcceptedTransaction`, but because of Rust's RAII
+    /// enforcement it has to be an array of options.  Otherwise we
+    /// would need to use unsafe to implement the fast pruning on the
+    /// array.
+    pub transaction_cache: Vec<Option<VersionedAcceptedTransaction>>,
 }
 
-#[async_trait::async_trait]
-impl<G: GenesisNetworkTrait, F: FaultInjection> Handler<InvalidatedBlockHashes>
-    for SumeragiWithFault<G, F>
-{
-    type Result = Vec<HashOf<VersionedValidBlock>>;
-
-    async fn handle(&mut self, InvalidatedBlockHashes: InvalidatedBlockHashes) -> Self::Result {
-        self.invalidated_blocks_hashes.clone()
+impl<F: FaultInjection> SumeragiWithFault<F> {
+    /// Get the current online peers by public key.
+    #[allow(clippy::expect_used)]
+    pub fn get_online_peer_keys(&self) -> Vec<PublicKey> {
+        self.current_online_peers
+            .lock()
+            .expect("lock on online peers")
+            .clone()
+            .into_iter()
+            .map(|peer_id| peer_id.public_key)
+            .collect()
     }
-}
-
-#[async_trait::async_trait]
-impl<G: GenesisNetworkTrait, F: FaultInjection> ContextHandler<Message>
-    for SumeragiWithFault<G, F>
-{
-    type Result = ();
 
-    async fn handle(&mut self, ctx: &mut Context<Self>, msg: Message) {
-        trace!(peer_role=?self.topology.role(&self.peer_id), ?msg);
-        if let Err(error) = msg.handle(self, ctx).await {
-            error!(%error, "Failed to handle message");
-        }
+    /// Set the public block hash to zero, in a thread-safe manner
+    #[allow(clippy::expect_used)]
+    pub fn zeroize(&self) {
+        *self
+            .latest_block_hash_for_use_by_block_sync
+            .lock()
+            .expect("Poisoned Mutex") = Hash::zeroed().typed()
     }
-}
-
-#[async_trait::async_trait]
-impl<G: GenesisNetworkTrait, F: FaultInjection> ContextHandler<RetrieveTransactions>
-    for SumeragiWithFault<G, F>
-{
-    type Result = ();
 
-    async fn handle(
-        &mut self,
-        ctx: &mut Context<Self>,
-        RetrieveTransactions: RetrieveTransactions,
-    ) {
-        if self.voting_in_progress().await {
-            return;
-        }
-        let txs = self.queue.get_transactions_for_block();
-        // TODO: This should properly process triggers
-        let event_recommendations = Vec::new();
-        if let Err(error) = self.round(txs, event_recommendations, ctx).await {
-            error!(%error, "Round failed");
+    /// Updates network topology by taking the actual list of peers from `WorldStateView`.
+    /// Updates it only if there is a change in WSV peers, otherwise leaves the order unchanged.
+    #[allow(clippy::expect_used)]
+    pub fn update_network_topology(topology: &mut Topology, wsv: &WorldStateView) {
```

Comment:
Could be removed, seems to pop up after merge.

Reply:
I'd keep it.

Reply:
But why? it's now `Topology` method, this function don't called anywhere, you could check this by removing `pub`.